### PR TITLE
Algod: fix nil deref while fetching stateproof secrets

### PR DESCRIPTION
**Pull request by algonathan**

### CODE REVIEW

**Comment by cce** [comment]
LGTM, would be great to get a test together that reproduces the issue and verifies it has been fixed now

**Review by cce — APPROVED** [review]

File: data/account/participationRegistry_test.go [hunk 1]
```diff
@@ -958,6 +957,36 @@ func TestAddStateProofKeys(t *testing.T) {
 	}
 }
 
+func TestGetRoundSecretsWhithout(t *testing.T) {
```
Comment (algoidan) [inline data/account/participationRegistry_test.go:960]: let's give it a meaningful name

File: data/accountManager_test.go [hunk 2]
```diff
@@ -248,3 +249,46 @@ func TestAccountManagerOverlappingStateProofKeys(t *testing.T) {
 	res = acctManager.StateProofKeys(basics.Round(merklesignature.KeyLifetimeDefault * 3))
 	a.Equal(1, len(res))
 }
+
+func TestAccountManagerOverlappingStateProofKeys1(t *testing.T) {
```
Comment (algoidan) [inline data/accountManager_test.go:253]: let's give it a meaningful name

File: data/account/participationRegistry.go [hunk 3]
```diff
@@ -220,6 +220,9 @@ var ErrNoKeyForID = errors.New("no valid key found for the participationID")
 // ErrSecretNotFound is used when attempting to lookup secrets for a particular round.
 var ErrSecretNotFound = errors.New("the participation ID did not have secrets for the requested round")
 
+// ErrStateProofVerifierIsNil states that no state proof field was found.
+var ErrStateProofVerifierIsNil = errors.New("record contains no StateProofVerifier")
```
Comment (winder) [inline data/account/participationRegistry.go:224]: nit:
```suggestion
// ErrStateProofVerifierNotFound states that no state proof field was found.
var ErrStateProofVerifierNotFound = errors.New("record contains no StateProofVerifier")
```

File: data/account/participationRegistry.go [hunk 4]
```diff
@@ -767,6 +770,9 @@ func (db *participationDB) GetStateProofSecretsForRound(id ParticipationID, roun
 	if err != nil {
 		return StateProofSecretsForRound{}, err
 	}
+	if partRecord.StateProof == nil {
+		return StateProofSecretsForRound{}, ErrStateProofVerifierIsNil
```
Comment (winder) [inline data/account/participationRegistry.go:774]: Provide the ID for debugging:
```suggestion
		return StateProofSecretsForRound{}, fmt.Errorf("%s: for participation ID %s", ErrStateProofVerifierNotFound, id)
```

**Review by winder — COMMENTED** [review]
Looks good, just a couple minor suggestions.

**Review by cce — APPROVED** [review]

File: data/accountManager.go [hunk 5]
```diff
@@ -79,10 +79,10 @@ func (manager *AccountManager) Keys(rnd basics.Round) (out []account.Participati
 // StateProofKeys returns a list of Participation accounts, and their stateproof secrets
 func (manager *AccountManager) StateProofKeys(rnd basics.Round) (out []account.StateProofSecretsForRound) {
 	for _, part := range manager.registry.GetAll() {
-		if part.OverlapsInterval(rnd, rnd) {
+		if part.StateProof != nil && part.OverlapsInterval(rnd, rnd) {
 			partRndSecrets, err := manager.registry.GetStateProofSecretsForRound(part.ParticipationID, rnd)
 			if err != nil {
-				manager.log.Errorf("error while loading round secrets from participation registry: %w", err)
```
Comment (algorandskiy) [inline data/accountManager.go:85]: why removed the err wrapping?

Comment (algoidan) [reply data/accountManager.go:85]: It is due to the fact that we don't return a wrapped error here, just log the message. 
Current, the log will contain a line like this:
```
error while loading round secrets from participation registry: %!w(*fmt.wrapError=&{...})
```

Comment (algorandskiy) [reply data/accountManager.go:85]: ah, right, logging, thanks!

**Review by algorandskiy — APPROVED** [review]

**Comment by algoidan** [comment]
Should the error be templated as %w?

**Pull request merged**